### PR TITLE
7.3 - Fix SecurityException fetching external media

### LINKED ISSUE #5818: Crash report 7.2.1: SecurityException: Permission Denial: com.google.android.apps.photos.contentprovider.MediaContentProvider
http://crashes.to/s/b4e0c187f19

```
Fatal Exception: java.lang.RuntimeException: An error occured while executing doInBackground()
       at android.os.AsyncTask$3.done(AsyncTask.java:300)
       at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:355)
       at java.util.concurrent.FutureTask.setException(FutureTask.java:222)
       at java.util.concurrent.FutureTask.run(FutureTask.java:242)
       at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:231)
       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112)
       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
       at java.lang.Thread.run(Thread.java:818)
Caused by java.lang.SecurityException: Permission Denial: reading com.google.android.apps.photos.contentprovider.MediaContentProvider uri content://com.google.android.apps.photos.contentprovider/0/1/mediakey%3A%2Flocal%253A7c1d3c4c-bfed-445c-a330-05ccafc1ca8f/ORIGINAL/NONE/1120761078 from pid=18359, uid=10262 requires the provider be exported, or grantUriPermission()
       at android.os.Parcel.readException(Parcel.java:1547)
       at android.database.DatabaseUtils.readExceptionFromParcel(DatabaseUtils.java:185)
       at android.database.DatabaseUtils.readExceptionWithFileNotFoundExceptionFromParcel(DatabaseUtils.java:148)
       at android.content.ContentProviderProxy.openTypedAssetFile(ContentProviderNative.java:689)
       at android.content.ContentResolver.openTypedAssetFileDescriptor(ContentResolver.java:1080)
       at android.content.ContentResolver.openAssetFileDescriptor(ContentResolver.java:921)
       at android.content.ContentResolver.openInputStream(ContentResolver.java:646)
       at com.yalantis.ucrop.task.BitmapLoadTask.copyFile(BitmapLoadTask.java:198)
       at com.yalantis.ucrop.task.BitmapLoadTask.processInputUri(BitmapLoadTask.java:176)
       at com.yalantis.ucrop.task.BitmapLoadTask.doInBackground(BitmapLoadTask.java:90)
       at com.yalantis.ucrop.task.BitmapLoadTask.doInBackground(BitmapLoadTask.java:41)
       at android.os.AsyncTask$2.call(AsyncTask.java:288)
       at java.util.concurrent.FutureTask.run(FutureTask.java:237)
       at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:231)
       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112)
       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
       at java.lang.Thread.run(Thread.java:818)
```

### PR DESCRIPTION
**Pull request by daniloercoli**
Fixes #5818 and #5672 by downloading the selected media file "in-line" in `onActivityResult` method.

**Details**
It seems that we're are using AsyncTask to download the content of the file from external app/content provider. Unfortunately this could lead to security errors due to design constraints.
`For security reasons, the permissions are temporary, so once the client app's task stack is finished, the file is no longer accessible. You must get the file data when you receive the intent answer, in the onActivityResult method. Store a copy of the file data, because the file won't be available anymore when onActivityResult returns.`

See `Access the Requested File` section here https://developer.android.com/training/secure-file-sharing/request-file.html#OpenFile

Also http://stackoverflow.com/questions/30572261/using-data-from-context-providers-or-requesting-google-photos-read-permission

### CODE REVIEW

**Comment by mzorz** [comment]
I'll review once conflicts are solved @daniloercoli 👍

**Comment by daniloercoli** [comment]
Merged `7.3` here, and fixed conflicts.

File: WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java [hunk 1]
```diff
@@ -1883,23 +1883,14 @@ public void onActivityResult(int requestCode, int resultCode, Intent data) {
...
-                    if (uri != null) {
-                        addMedia(uri);
-                    } else {
-                        Toast.makeText(EditPostActivity.this, getString(R.string.error_downloading_image),
-                                Toast.LENGTH_SHORT).show();
-                    }
-                }
-            }.executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR, mediaUri);
+            // Do not download the file in async task. See https://github.com/wordpress-mobile/WordPress-Android/issues/5818
+            Uri downloadedUri = MediaUtils.downloadExternalMedia(EditPostActivity.this, mediaUri);
```
Comment (mzorz) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java]: should we enclose this call in the `try...catch(IllegalStateException)` block as you've done in this line https://github.com/wordpress-mobile/WordPress-Android/pull/5827/files#diff-fdc9b5c706ac889526b8df93c55351ceR669 ?

**Comment by mzorz** [comment]
Just added one comment - let me know what you think and we're good to go @daniloercoli

File: WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java (hunk 1 above)
Comment (daniloercoli) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java]: In theory yes, but we didn't get reports about it here. Anyway, I can add `try/catch IllegalStateException` here, and see if we get reports. Both of those should be removed soon.

**Comment by mzorz** [comment]
awesome - waiting for Travis to do its thing then we'll :shipit:

**Comment by mzorz** [comment]
:shipit:

**Pull request merged**